### PR TITLE
ARROW-15026: [Python] Error if datetime.timedelta to pyarrow.duration conversion overflows

### PR DESCRIPTION
**Pull request by anjakefala**

### CODE REVIEW

File: cpp/src/arrow/python/datetime.h [hunk 1]
```diff
@@ -35,6 +37,8 @@
 #define PyDateTimeAPI ::arrow::py::internal::datetime_api
 
 namespace arrow {
+using internal::MultiplyWithOverflow;
+using internal::AddWithOverflow;
```
Comment (jorisvandenbossche) [inline cpp/src/arrow/python/datetime.h:41]: Either you can leave this out, or either you don't have to include the `arrow::internal::` below when using those, I think

File: cpp/src/arrow/python/datetime.h [hunk 2]
```diff
@@ -147,14 +151,25 @@ inline int64_t PyDelta_to_ms(PyDateTime_Delta* pytimedelta) {
 }
 
 ARROW_PYTHON_EXPORT
-inline int64_t PyDelta_to_us(PyDateTime_Delta* pytimedelta) {
-  return (PyDelta_to_s(pytimedelta) * 1000000LL +
-          PyDateTime_DELTA_GET_MICROSECONDS(pytimedelta));
+// TODO which PyDeltas do we cover?
```
Comment (jorisvandenbossche) [inline cpp/src/arrow/python/datetime.h:154]: What's the question exactly?

File: python/pyarrow/tests/test_types.py [hunk 3]
```diff
@@ -845,6 +845,11 @@ def test_decimal_overflow():
         with pytest.raises(ValueError):
             pa.decimal256(i, 0)
 
+def test_timedelta_overflow()::
+    d = datetime.timedelta(days=-106751992, seconds=71945, microseconds=224192)
+    with pytest.raises(pa.ArrowInvalid):
+        pa.scalar(d)
```
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_types.py:851]: The current test looks good, but will only test it for a single resolution (the default, which is micro, I think), and thus not cover all your changes. So maybe add some additional cases where you pass the data type (with specific resolution) explicitly)

**Review by jorisvandenbossche — COMMENTED** [review]
Thanks a lot for the PR!

Question: you only added overflow checks for micro and nano resolutions. Do we need that for second and milli as well, or will those never run into an overflow? (since a python timedelta also stores days, I would assume also second can still run into an overflow)

**Comment by notEvil** [comment]
Thanks for taking over!

> Question: you only added overflow checks for micro and nano resolutions. Do we need that for second and milli as well, or will those never run into an overflow? (since a python timedelta also stores days, I would assume also second can still run into an overflow)

Both `datetime.timedelta.min` and `datetime.timedelta.max` fit into int64 at s and ms resolution.

File: cpp/src/arrow/python/datetime.h (hunk 2 above)
Comment (notEvil) [reply cpp/src/arrow/python/datetime.h:154]: I'd assume the min/max value of `datetime.timedelta` for us and ns. For us its `datetime.timedelta(days=-106751992, seconds=71945, microseconds=224192)` to `datetime.timedelta(days=106751991, seconds=14454, microseconds=775807)` and if i'm not miscalculating `datetime.timedelta(days=-106752, seconds=763, microseconds=145224)` to `datetime.timedelta(days=106751, seconds=85636, microseconds=854775)` for ns

**Comment by jorisvandenbossche** [comment]
> Both `datetime.timedelta.min` and `datetime.timedelta.max` fit into int64 at s and ms resolution.

Ah, indeed, the days component being limited to 999999999 (and not just any int) means that this will always fit for s/ms

**Comment by jorisvandenbossche** [comment]
@anjakefala small workflow question: can you add new changes after a review as a new commit instead of combining it in the original commit? That makes it easier to see what changes since the last review.

**Comment by anjakefala** [comment]
@jorisvandenbossche Oh! Yes, of course! I find it easier to rebase along the way, instead of needing to do one messy rebase in the end, if I have more than one unit of work in the PR. This PR is only a single unit of work, so it is straightforward to just add commits.
Screenshot from 2022-08-09 16-19-58

I do not mind making this workflow shift, but in the GitHub UI, you can click compare on the Force Push to see what the difference is. 

Is it generally the process to add messy commits, and then everything gets squashed into a single commit in the end?

**Comment by anjakefala** [comment]
I am trying to set up clang linting in my local enviornment for Arrow, which is why this PR is currently quiet.

File: python/pyarrow/tests/test_types.py (hunk 3 above)
Comment (anjakefala) [reply python/pyarrow/tests/test_types.py:851]: I agree it is a good idea to add test cases!

I think I am missing something, but it seems like datetime.timedelta is only at the microsecond resolution: https://docs.python.org/3/library/datetime.html#available-types

Comment (anjakefala) [reply python/pyarrow/tests/test_types.py:851]: Nvm, I think I figured it out!

Comment (anjakefala) [reply python/pyarrow/tests/test_types.py:864]: Added tests for additional resolutions

File: cpp/src/arrow/python/datetime.h (hunk 2 above)
Comment (anjakefala) [reply cpp/src/arrow/python/datetime.h]: Yup! I have also confirmed that at int64, seconds and milliseconds fit without overflow.

File: python/pyarrow/src/arrow/python/datetime.h [hunk 4]
```diff
@@ -20,9 +20,13 @@
 #include <algorithm>
 #include <chrono>
 
+#include "arrow/python/platform.h"
+#include "arrow/python/visibility.h"
```
Comment (jorisvandenbossche) [inline python/pyarrow/src/arrow/python/datetime.h:24]: I suppose this is a left-over from rebasing on top of the moved code (those lines can be removed, they are already included a bit below as well)

File: python/pyarrow/tests/test_types.py [hunk 5]
```diff
@@ -857,6 +857,31 @@ def test_decimal_overflow():
...
 
 
+def test_timedelta_overflow():
+    # microsecond resolution, overflow
+    d = datetime.timedelta(days=-106751992, seconds=71945, microseconds=224192)
+    with pytest.raises(pa.ArrowInvalid):
+        pa.scalar(d)
+
+    # microsecond resolution, overflow
+    d = datetime.timedelta(days=106751992, seconds=14454, microseconds=775807)
```
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_types.py:867]: This one will overflow based on the days alone. We can also add a case that will only overflow when adding the microseconds. I think `datetime.timedelta(days=106751991, seconds=14454, microseconds=775808)` should to the trick (`datetime.timedelta(days=106751991, seconds=14454, microseconds=775807)` is the limit that works, so adding one more to the microseconds)

**Review by jorisvandenbossche — COMMENTED** [review]
Looking good! Added two minor comments

**Review by wjones127 — APPROVED** [review]
This looks good, and it seems you have addressed @jorisvandenbossche 's final comments. I will merge.

**Comment by anjakefala** [comment]
Thank you @jorisvandenbossche and @wjones127 and @notEvil!

**Pull request merged**